C7GameData/AIData/TileKnowledge.cs

```diff
@@ -0,0 +1,34 @@
+using System.Collections.Generic;
+
+namespace C7GameData
+{
+	public class TileKnowledge
+	{
+		HashSet<Tile> knownTiles = new HashSet<Tile>();
+		HashSet<Tile> visibleTiles = new HashSet<Tile>();
+
+		public void AddTilesToKnown(Tile unitLocation) {
+			knownTiles.Add(unitLocation);
+			foreach (Tile t in unitLocation.neighbors.Values) {
+				knownTiles.Add(t);
+			}
+		}
+
+		public bool isTileKnown(Tile t) {
+			return knownTiles.Contains(t);
+		}
+
+		/**
+		 * Returns a copy of the list of known tiles.
+		 * This prevents external modifications.
+		 **/
+		public List<Tile> AllKnownTiles()
```

Comment:
is there a particular reason for returning the copy as a list?

Reply:
Potentially excessive caution around the data being modified outside of the TileKnowledge API.

The gist of the thinking was that there are places where it's handy for the AI to have a list of all known tiles - for example, deciding where to send a settler.  But it should never be possible to remove items from that list, and if you send the whole list out, it is.

In Java, I would have done this via the [Collections.unmodifiableList](https://docs.oracle.com/javase/1.5.0/docs/api/java/util/Collections.html#unmodifiableList(java.util.List)) method, which returns a read-only view of the list.  This is lighter weight than making a whole copy, but I'm not aware of an equivalent in C#.

Reply:
Yeah it makes senes to to return a copy, I was just curious if there's a reason why the "real" known tiles is in a map, but the AllKnownTiles returns a copy as a list, as opposed to a copy as a map.

Reply:
Oh, right.

The internal collection is a Set so that if a tile is added more than once via _AddTilesToKnown_, it still only winds up there once.  My intention with _AddTilesToKnown_ is that by passing it the tile a unit is on, it can automatically handle logic around how many neighboring tiles should be added, including based on terrain height (hills, mountains, etc.).  A set seemed like the simplest way to avoid adding duplicates.

For consumption purposes, a list just seemed like the default collection.  A set would work just as well, though.

I'm guessing you meant "set" instead of "map" in both cases?

Reply:
_Long_ term, we'll also need more methods to deal with tiles revealed by air missions... but that can probably just be a new method similar to AddTilesToKnown, but that doesn't take into account ground terrain, and has a wider radius.  I think it will be compatible, but am not too concerned about it now.